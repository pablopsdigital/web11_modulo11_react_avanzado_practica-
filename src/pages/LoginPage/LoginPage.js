import React, { useEffect, useState } from 'react';
import Button from '../../components/Button/Button';
import './LoginPage.scss';
import { ReactComponent as Brand } from '../../images/svg/brand.svg';
import Alert from '../../components/Alert/Alert';
import SpinnerLoading from '../../components/SpinnerLoading/SpinnerLoading';
import { useDispatch, useSelector } from 'react-redux';
import { authLogin, resetError } from '../../redux/actions';
import { getIsLogged, getUi } from '../../redux/selectors';

function LoginPage({ history, ...props }) {
  const auth = useSelector(getIsLogged);
  const { isLoading, error } = useSelector(getUi);

  useEffect(() => {
    if (auth === true) {
      history.push('/adverts');
    }
  }, [auth, history]);

  //State for inputs values
  const [inputsValuesState, setInputsValue] = useState({
    email: '',
    password: '',
    rememberme: false
  });

  const handleChange = (event) => {
    const input = event.target;
    const value = input.type === 'checkbox' ? input.checked : input.value;
    const name = input.name;

    setInputsValue((prevState) => ({
      ...prevState,
      [name]: value
    }));
  };

  const dispatch = useDispatch();
  const handleFormSubmit = async (event) => {
    event.preventDefault();
    dispatch(authLogin(inputsValuesState));
  };

  return (
    <div id="loginPage">
      <div className="brand-container">
        <Brand className="brand-img" alt="brand" />
      </div>
      <div className="content">
        <h1>Login</h1>
        <form onSubmit={handleFormSubmit}>
          <input
            type="email"
            name="email"
            placeholder="Enter email"
            value={inputsValuesState.email}
            onChange={handleChange}
          />
          <br />
          <input
            type="password"
            name="password"
            placeholder="Enter password"
            value={inputsValuesState.password}
            onChange={handleChange}
          />
          <br />
          <label>
            <input
              type="checkbox"
              name="rememberme"
              checked={inputsValuesState.rememberme}
              onChange={handleChange}
            />{' '}
            Remenber me
          </label>
          <br />
          <Button
            type="submit"
            disabled={isLoading || !inputsValuesState.email || !inputsValuesState.password}
          >
            Iniciar sesi√≥n
          </Button>
        </form>
      </div>
      {isLoading && <SpinnerLoading />}
      {error && <Alert className="loginPage-error">{error.message}</Alert>}
    </div>
  );
}

export default LoginPage;
