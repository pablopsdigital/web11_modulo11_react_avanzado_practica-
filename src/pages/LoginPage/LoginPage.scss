#loginPage {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 100vh;
  .brand-container {
    margin-bottom: 2rem;
  }
  .content {
    background-color: white;
    padding: 2rem 3rem;
    border-radius: 0.5rem;
    border: 1px solid #c2c5e1;
  }
}
