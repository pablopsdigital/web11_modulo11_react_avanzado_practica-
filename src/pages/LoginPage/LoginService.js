import ApiClient, {
  removeAuthorizationHeader,
  setAuthorizationHeader
} from '../../services/ApiClient';
import CustomLocalStorageManager from '../../utils/CustomLocalStorageManager';

export const login = (credentials) => {
  return (
    ApiClient.post('/api/auth/login', credentials)
      //Save token in localStorage and default header axios
      .then(({ accessToken }) => {
        setAuthorizationHeader(accessToken);
        if (credentials.rememberme) {
          CustomLocalStorageManager.setItem('token', accessToken);
        }
        return accessToken;
      })
  );
};

export const logout = () =>
  Promise.resolve().then(() => {
    //Delete headers autentication axios and delete auth of localStorage
    removeAuthorizationHeader();
    //CustomLocalStorageManager.clear();
    CustomLocalStorageManager.removeItem('token');
  });
