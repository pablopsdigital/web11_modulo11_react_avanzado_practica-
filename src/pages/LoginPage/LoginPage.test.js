import { fireEvent, render } from '@testing-library/react';
import { LoginPage } from './LoginPage';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import React from 'react';

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
  useState: jest.fn()
}));

const isLoading = jest.fn();
const error = jest.fn();

describe('LoginPage', () => {
  const mockStore = {
    getState: () => ({
      auth: false,
      adverts: {
        loaded: false,
        data: [],
        tags: []
      },
      ui: {
        isLoading: false,
        error: null
      },
      filters: {}
    }),
    subscribe: () => {},
    dispatch: () => {}
  };
  // const authLogin = jest.fn().mockResolvedValue();
  const email = 'username';
  const password = '1234';

  test('snapshot', () => {
    const { container } = render(<LoginPage />);
    expect(container).toMatchSnapshot();
  });

  test('should call authLogin', () => {
    const { getByLabelText, getByRole, debug } = render(
      <Provider store={mockStore}>
        <BrowserRouter>
          <LoginPage></LoginPage>
        </BrowserRouter>
      </Provider>
    );
    //Select elements
    const emailInput = getByLabelText(/email/);
    const passwordInput = getByLabelText(/password/);
    const remembermeInput = getByLabelText(/rememberme/);
    const buttonSubmit = getByRole('button');
    //Send event
    fireEvent.change(emailInput, { target: { value: email } });
    fireEvent.change(passwordInput, { target: { value: password } });
    fireEvent.change(remembermeInput, { target: { value: true } });
    fireEvent.click(buttonSubmit);
    //Check Result
    expect(authLogin).toHaveBeenCalledWith({ email, password });
  });

  // test('should reset error', () => {
  //   const error = { message: 'This is a error' };
  //   const OnResetError = jest.fn();
  //   const { getByText } = render(<LoginPage error={error} onResetError={OnResetError} />);
  //   const errorField = getByText(error.message);
  //   fireEvent.click(errorField);
  //   expect(OnResetError).toHaveBeenCalled();
  // });
});
