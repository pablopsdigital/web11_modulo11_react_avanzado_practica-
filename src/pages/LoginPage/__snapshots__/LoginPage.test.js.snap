// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage snapshot 1`] = `
<div>
  <div
    id="loginPage"
  >
    <div
      class="brand-container"
    >
      <svg
        alt="brand"
        class="brand-img"
      >
        brand.svg
      </svg>
    </div>
    <div
      class="content"
    >
      <h1>
        Login
      </h1>
      <form>
        <input
          name="email"
          placeholder="Enter email"
          type="email"
          value=""
        />
        <br />
        <input
          name="password"
          placeholder="Enter password"
          type="password"
          value=""
        />
        <br />
        <label>
          <input
            name="rememberme"
            type="checkbox"
          />
           
          Remenber me
        </label>
        <br />
        <button
          disabled=""
          type="submit"
        >
          Iniciar sesi√≥n
        </button>
      </form>
    </div>
  </div>
</div>
`;
